src/blob_file_iterator.cc

```diff
@@ -170,45 +169,51 @@ void BlobFileIterator::GetBlobRecord() {
         decoder_.DecodeRecord(&record_slice, &cur_blob_record_, &uncompressed_,
                               titan_cf_options_.memory_allocator());
   }
-  if (!status_.ok()) return;
+  if (!status_.ok()) return false;
 
   cur_record_offset_ = iterate_offset_;
   cur_record_size_ = kRecordHeaderSize + record_size;
   iterate_offset_ += cur_record_size_;
-  AdjustOffsetToNextBlockHead();
+  if (block_size_ != 0) iterate_offset_ = Roundup(iterate_offset_, block_size_);
   valid_ = true;
+  return true;
 }
 
 void BlobFileIterator::PrefetchAndGet() {
-  if (iterate_offset_ >= end_of_blob_record_) {
-    valid_ = false;
-    return;
-  }
+  while (iterate_offset_ < end_of_blob_record_) {
+    if (readahead_begin_offset_ > iterate_offset_ ||
+        readahead_end_offset_ < iterate_offset_) {
+      // alignment
+      readahead_begin_offset_ =
+          iterate_offset_ - (iterate_offset_ & (kDefaultPageSize - 1));
+      readahead_end_offset_ = readahead_begin_offset_;
+      readahead_size_ = kMinReadaheadSize;
+    }
+    auto min_blob_size =
+        iterate_offset_ + kRecordHeaderSize + titan_cf_options_.min_blob_size;
+    if (readahead_end_offset_ <= min_blob_size) {
+      while (readahead_end_offset_ + readahead_size_ <= min_blob_size &&
+             readahead_size_ < kMaxReadaheadSize)
+        readahead_size_ <<= 1;
+      file_->Prefetch(readahead_end_offset_, readahead_size_);
+      readahead_end_offset_ += readahead_size_;
+      readahead_size_ = std::min(kMaxReadaheadSize, readahead_size_ << 1);
+    }
 
-  if (readahead_begin_offset_ > iterate_offset_ ||
-      readahead_end_offset_ < iterate_offset_) {
-    // alignment
-    readahead_begin_offset_ =
-        iterate_offset_ - (iterate_offset_ & (kDefaultPageSize - 1));
-    readahead_end_offset_ = readahead_begin_offset_;
-    readahead_size_ = kMinReadaheadSize;
-  }
-  auto min_blob_size =
-      iterate_offset_ + kRecordHeaderSize + titan_cf_options_.min_blob_size;
-  if (readahead_end_offset_ <= min_blob_size) {
-    while (readahead_end_offset_ + readahead_size_ <= min_blob_size &&
-           readahead_size_ < kMaxReadaheadSize)
-      readahead_size_ <<= 1;
-    file_->Prefetch(readahead_end_offset_, readahead_size_);
-    readahead_end_offset_ += readahead_size_;
-    readahead_size_ = std::min(kMaxReadaheadSize, readahead_size_ << 1);
-  }
+    bool live = GetBlobRecord();
 
-  GetBlobRecord();
+    if (readahead_end_offset_ < iterate_offset_) {
+      readahead_end_offset_ = iterate_offset_;
+    }
 
-  if (readahead_end_offset_ < iterate_offset_) {
-    readahead_end_offset_ = iterate_offset_;
+    // If the record is a hole-punch record, we should continue to the next
+    // record by adjusting iterate_offset_, otherwise (not a hole-punch record),
+    // we should break the loop and return the record, iterate_offset_ is
+    // already adjusted inside GetBlobRecord() in this case.
+    if (live || !status().ok()) return;
```

Comment:
For example, IO errors, In this case, we should not continue the loop.

As for whether setting `valid_` to false, I don't think it is necessary, since `valid()` (the function not the variable) considers `status_` already. So I didn't change the behavior (current code does not set `valid_` either when encountering io errors).